### DISCUSSION TITLE
Can't compile ContentView pages with data bindings displayed on MainPage

### DISCUSSION THREAD

**Discussion by King-Taz** [dotnet/maui#26797] [General] [answered]
Beginner here, asking for help or advice. I have two ContentView pages (that have data binding, using MVVM toolkit) that are added to the MainPage as below. 

![image](https://github.com/user-attachments/assets/5a5f34c8-1735-46d2-8817-33860f2de1ff)

Everything works well, but I don't understand how to reference the DataType so that the bindings will compile. Setting the DataType to what was used for the ContentView pages 

```x:DataType="viewmodel:SharedViewModel"```

doesn't correct the issue. I've tried everything I could find, any ideas? Can the debugger be used to discover how the system resolved the dependencies at run-time? I could be adding the ContentView pages incorrectly, I don't know.

**Comment by snpkhy** [comment]
> Beginner here, asking for help or advice. I have two ContentView pages (that have data binding, using MVVM toolkit) that are added to the MainPage as below.
> 
> ![image](https://private-user-images.githubusercontent.com/148505034/398305274-5a5f34c8-1735-46d2-8817-33860f2de1ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUwNTE1NTksIm5iZiI6MTczNTA1MTI1OSwicGF0aCI6Ii8xNDg1MDUwMzQvMzk4MzA1Mjc0LTVhNWYzNGM4LTE3MzUtNDZkMi04ODE3LTMzODYwZjJkZTFmZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNFQxNDQwNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYzQyOWZiZWRjZjBmMTM3Nzc1MWExMzAxZjU5OTc5ODFhZmE4NTkxNTNhMDQwYjlhM2MyNjFjZmUxNDBjNjgzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ROruk9njUzSm6eaxUOMsRcbv7hIsol7vp3xYQomLEzQ)
> 
> Everything works well, but I don't understand how to reference the DataType so that the bindings will compile. Setting the DataType to what was used for the ContentView pages
> 
> `x:DataType="viewmodel:SharedViewModel"`
> 
> doesn't correct the issue. I've tried everything I could find, any ideas? Can the debugger be used to discover how the system resolved the dependencies at run-time? I could be adding the ContentView pages incorrectly, I don't know.

Do you connect your pages in maui program?

**Reply by King-Taz** [reply]
Page1View and Page2View contents are displayed together on MainPage, if that's the question

**Reply by snpkhy** [reply]
> Page1View and Page2View contents are displayed together on MainPage, if that's the question

No, I mean u have to write services... 
        builder.Services.AddSingleton<MainPageViewModel>();
        builder.Services.AddSingleton<MainPage>();
u have to do it for all viewmodels u wanna to use and of course pages. it only just question if u do that. if yes i smth strange :)

**Reply by King-Taz** [reply]
Yes, I'm using dependency injection, but that doesn't help since it works equally well if I don't use those services.

**Reply by snpkhy** [reply]
> Yes, I'm using dependency injection, but that doesn't help since it works equally well if I don't use those services.

Do you bind your viewmodel with code behind in constructor? Did you try smth like this 
<ContentPage.BindingContext>
        <local:ClockViewModel />
    </ContentPage.BindingContext>

**Reply by King-Taz** [reply]
I tried this but it didn't solve the problem. Instancing the viewmodel as a resource and referencing as a StaticResource solved the issue.  There's probably other methods that work.

**Comment by dartasen** [comment]
* Your ContentPage (BindinApp.MainPage) should provide the x:DataType that corresponds to BindingContext, which in your case is what's holding SharedViewModel. It will the Xaml Compiler to look at given type, where he should find SharedViewModel, that will allow him to compiled the binding instead of using Reflection to find the property.
OR
* Instanciate your viewmodel in your <ContentPage.Resources> and give it to Page1View & Page2View

**Reply by King-Taz** [reply, marked as answer]
> Your ContentPage (BindinApp.MainPage) should provide the x:DataType that corresponds to BindingContext.

The app doesn't work if x:DataType is given on MainPage as was done on Page1View and Page2View.

> Instanciate your viewmodel in your <ContentPage.Resources> and give it to Page1View & Page2View

This worked, thanks! Implemented as follows (maybe there's a better way, I don't know?)

```
    <ContentPage.Resources>
        <ResourceDictionary>
            <local:SharedViewModel x:Key="SharedViewModel" />
        </ResourceDictionary>
    </ContentPage.Resources>

    <StackLayout>
        <local:Page1View BindingContext="{StaticResource SharedViewModel}"/>
        <local:Page2View BindingContext="{StaticResource SharedViewModel}"/>
    </StackLayout>
```

**Reply by dartasen** [reply]
There's more complex approach, but this one is the nice and simple
